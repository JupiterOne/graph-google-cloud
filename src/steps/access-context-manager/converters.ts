import { accesscontextmanager_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
  ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
  ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
  ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
  ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
  ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
} from './constants';

export function createAccessPolicyEntity(
  data: accesscontextmanager_v1.Schema$AccessPolicy,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
        _class: ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
        name: data.name,
        title: data.title,
        etag: data.etag,
      },
    },
  });
}

export function createAccessLevelEntity(
  data: accesscontextmanager_v1.Schema$AccessLevel,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
        _class: ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
        name: data.name,
        description: data.description,
        title: data.title,
      },
    },
  });
}

export function createServicePerimeterEntity(
  data: accesscontextmanager_v1.Schema$ServicePerimeter,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
        _class: ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
        name: data.name,
        description: data.description,
        perimeterType: data.perimeterType,
        title: data.title,
        useExplicitDryRunSpec: data.useExplicitDryRunSpec,
        // shared fields between .spec/.status depending on the run mode
        accessLevels: data.status?.accessLevels,
        resources: data.status?.resources,
        restrictedServices: data.status?.restrictedServices,
        // ingressPolicies
        // egressPolicies
      },
    },
  });
}
