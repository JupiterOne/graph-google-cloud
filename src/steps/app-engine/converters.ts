import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { appengine_v1 } from 'googleapis';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  ENTITY_CLASS_APP_ENGINE_APPLICATION,
  ENTITY_CLASS_APP_ENGINE_INSTANCE,
  ENTITY_CLASS_APP_ENGINE_SERVICE,
  ENTITY_CLASS_APP_ENGINE_VERSION,
  ENTITY_TYPE_APP_ENGINE_APPLICATION,
  ENTITY_TYPE_APP_ENGINE_INSTANCE,
  ENTITY_TYPE_APP_ENGINE_SERVICE,
  ENTITY_TYPE_APP_ENGINE_VERSION,
} from './constants';

export function createAppEngineApplicationEntity(
  data: appengine_v1.Schema$Application,
  projectId: string,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_APPLICATION,
        _type: ENTITY_TYPE_APP_ENGINE_APPLICATION,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        location: data.locationId,
        servingStatus: data.servingStatus,
        defaultHostname: data.defaultHostname,
        defaultBucket: data.defaultBucket,
        codeBucket: data.codeBucket,
        authDomain: data.authDomain,
        grcDomain: data.gcrDomain,
        splitHealthChecks: data.featureSettings?.splitHealthChecks,
        useContainerOptimizedOs: data.featureSettings?.useContainerOptimizedOs,
        webLink: getGoogleCloudConsoleWebLink(
          `/appengine?project=${projectId}`,
        ),
      },
    },
  });
}

export function createAppEngineServiceEntity(
  data: appengine_v1.Schema$Service,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_SERVICE,
        _type: ENTITY_TYPE_APP_ENGINE_SERVICE,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
      },
    },
  });
}

export function createAppEngineVersionEntity(
  data: appengine_v1.Schema$Version,
  projectId: string,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_VERSION,
        _type: ENTITY_TYPE_APP_ENGINE_VERSION,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        category: ['application'],
        function: ['workload-management'],
        // The following properties are shared (both standard and flexible)
        versionUrl: data.versionUrl,
        threadsafe: data.threadsafe,
        // Tells us if this version is using standard of flexible env
        env: data.env,
        runtime: data.runtime,
        servingStatus: data.servingStatus,
        createdBy: data.createdBy,
        // Standard specific properties
        instanceClass: data.instanceClass,
        diskUsageBytes:
          data.diskUsageBytes && parseInt(data.diskUsageBytes as string, 10),
        // Flexible specific properties
        manualScalingInstances: data.manualScaling?.instances,
        cpuCount: data.resources?.cpu,
        diskGb: data.resources?.diskGb,
        memoryGb: data.resources?.memoryGb,
        readinessCheckFailureThreshold: data.readinessCheck?.failureThreshold,
        readinessCheckSuccessThreshold: data.readinessCheck?.successThreshold,
        readinessCheckInterval: data.readinessCheck?.checkInterval,
        readinessCheckTimeout: data.readinessCheck?.timeout,
        readinessCheckAppStartTimeout: data.readinessCheck?.appStartTimeout,
        livenessCheckFailureThreshold: data.livenessCheck?.failureThreshold,
        livenessCheckSuccessThreshold: data.livenessCheck?.successThreshold,
        livenessCheckInterval: data.livenessCheck?.checkInterval,
        livenessCheckTimeout: data.livenessCheck?.timeout,
        livenessCheckInitialDelay: data.livenessCheck?.initialDelay,
        createdOn: parseTimePropertyValue(data.createTime),
        webLink: getGoogleCloudConsoleWebLink(
          `/appengine/versions?serviceId=${data.name?.split(
            '/',
          )[3]}&project=${projectId}`,
        ),
      },
    },
  });
}

export function createAppEngineInstanceEntity(
  data: appengine_v1.Schema$Instance,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_INSTANCE,
        _type: ENTITY_TYPE_APP_ENGINE_INSTANCE,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        // Standard specific properties
        appEngineRelease: data.appEngineRelease,
        availability: data.availability,
        requests: data.requests,
        averageLatency: data.averageLatency,
        // Turning this to int so that you can apply ASC,DESC inside J1QL on it
        memoryUsage:
          data.memoryUsage && parseInt(data.memoryUsage as string, 10),
        // Flexible specific properties
        vmStatus: data.vmStatus,
        vmIp: data.vmIp,
        vmLiveness: data.vmLiveness,
        createdOn: parseTimePropertyValue(data.startTime),
        hostname: null,
      },
    },
  });
}
