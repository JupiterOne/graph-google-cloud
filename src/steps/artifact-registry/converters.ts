import { artifactregistry_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  ARTIFACT_REGISTRY_CLASS,
  ARTIFACT_REGISTRY_REPOSITORY_CLASS,
  ARTIFACT_REGISTRY_REPOSITORY_TYPE,
  ARTIFACT_REGISTRY_TYPE,
  ARTIFACT_REGISTRY_VPCSC_CONFIGURATION_CLASS,
  ARTIFACT_REGISTRY_VPCSC_CONFIGURATION_TYPE,
  ARTIFACT_REGISTRY_VPCSC_POLICY_CLASS,
  ARTIFACT_REGISTRY_VPCSC_POLICY_TYPE,
  ARTIFACT_REPOSITORY_PACKAGE_CLASS,
  ARTIFACT_REPOSITORY_PACKAGE_TYPE,
} from './constants';

export function createArtifactRegistryRepositoryEntity(
  data: artifactregistry_v1.Schema$Repository,
  projectId: string,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ARTIFACT_REGISTRY_REPOSITORY_TYPE,
        _class: ARTIFACT_REGISTRY_REPOSITORY_CLASS,
        name: data.name,
        createdTime: data.createTime,
        updatedTime: data.updateTime,
        format: data.format,
        mode: data.mode,
        kmsKey: data.kmsKeyName,
      },
    },
  });
}

export function createArtifactRepositoryPackageEntity(
  data: artifactregistry_v1.Schema$Package,
  projectId: string,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ARTIFACT_REPOSITORY_PACKAGE_TYPE,
        _class: ARTIFACT_REPOSITORY_PACKAGE_CLASS,
        name: data.name,
        createdTime: data.createTime,
        updatedTime: data.updateTime,
        repositoryName: data.name as string,
      },
    },
  });
}

export function createArtifactRegistryEntity(
  organizationId: string,
  data: any,
  projectId: string,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: (organizationId + '_' + ARTIFACT_REGISTRY_TYPE) as string,
        _type: ARTIFACT_REGISTRY_TYPE,
        _class: ARTIFACT_REGISTRY_CLASS,
        name: 'Artifact Registry Service',
        function: ['container-registry'],
        category: ['software'],
        endpoint: 'http://console.cloud.google.com/',
      },
    },
  });
}

function getVpcscPolicyStatus(data) {
  if (
    data.spec.restrictedServices.includes('artifactregistry.googleapis.com')
  ) {
    return 'DENY';
  } else if (
    data.spec?.vpcAccessibleServices?.allowedServices?.includes(
      'artifactregistry.googleapis.com',
    )
  ) {
    return 'ALLOW';
  } else {
    return 'VPCSC_POLICY_UNSPECIFIED';
  }
}

export function createArtifactRegistryVpcscConfigEntity(
  data,
  projectId: string,
) {
  const vpcPolicyStatus = getVpcscPolicyStatus(data);
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: `Artifact_VPC_SC_CONFIG:${data.title}`,
        _type: ARTIFACT_REGISTRY_VPCSC_CONFIGURATION_TYPE,
        _class: ARTIFACT_REGISTRY_VPCSC_CONFIGURATION_CLASS,
        name: data.title,
        VPCSCPolicy: vpcPolicyStatus,
        projectId: projectId,
      },
    },
  });
}

export function createArtifactRegistryVpcPolicyEntity(data, projectId: string) {
  const vpcPolicyStatus = getVpcscPolicyStatus(data);
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: `Artifact_VPC_SC_Policy:${data.name.split('/')[1]}`,
        _type: ARTIFACT_REGISTRY_VPCSC_POLICY_TYPE,
        _class: ARTIFACT_REGISTRY_VPCSC_POLICY_CLASS,
        name: `Artifact Policy: ${data.name.split('/')[1]}`,
        VPCSCPolicy: vpcPolicyStatus,
        projectId: projectId,
      },
    },
  });
}
