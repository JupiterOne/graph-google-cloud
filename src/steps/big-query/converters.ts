import { bigquery_v2 } from 'googleapis';
import {
  GoogleCloudResourceData,
  createGoogleCloudIntegrationEntity,
} from '../../utils/entity';
import {
  BIG_QUERY_DATASET_ENTITY_CLASS,
  BIG_QUERY_DATASET_ENTITY_TYPE,
  BIG_QUERY_TABLE_ENTITY_CLASS,
  BIG_QUERY_TABLE_ENTITY_TYPE,
  BIG_QUERY_MODEL_ENTITY_TYPE,
  BIG_QUERY_MODEL_ENTITY_CLASS,
} from './constants';
import { isMemberPublic } from '../../utils/iam';
import { BigQueryTable } from './client';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';

interface DatasetAccess {
  domain?: string;
  groupByEmail?: string;
  iamMember?: string;
  role?: string;
  specialGroup?: string;
  userByEmail?: string;
}

// 7.1 Ensure that BigQuery datasets are not anonymously or publicly accessible (Scored)
function isBigQueryDatasetPublicAccess(accessList: DatasetAccess[]): boolean {
  for (const access of accessList) {
    const value =
      access.groupByEmail ||
      access.userByEmail ||
      access.iamMember ||
      access.specialGroup ||
      '';
    if (isMemberPublic(value)) {
      return true;
    }
  }

  return false;
}

export function createBigQueryDatasetEntity(data: bigquery_v2.Schema$Dataset) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data as GoogleCloudResourceData,
      assign: {
        _key: data.id as string,
        _type: BIG_QUERY_DATASET_ENTITY_TYPE,
        _class: BIG_QUERY_DATASET_ENTITY_CLASS,
        name: data.datasetReference?.datasetId,
        description: data.description,
        public: isBigQueryDatasetPublicAccess(data.access || []),
        // 7.3 Ensure that a Default Customer-managed encryption key (CMEK) is specified for all BigQuery Data Sets
        kmsKeyName: data.defaultEncryptionConfiguration?.kmsKeyName,
        location: data.location,
        encrypted: true,
        classification: null,
        etag: data.etag,
        createdOn: data.creationTime
          ? parseInt(data.creationTime, 10)
          : undefined,
        updatedOn: data.lastModifiedTime
          ? parseInt(data.lastModifiedTime, 10)
          : undefined,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigquery?project=${data.datasetReference?.projectId}&d=${data.datasetReference?.datasetId}&p=${data.datasetReference?.projectId}&page=dataset`,
        ),
      },
    },
  });
}

function getBigQueryModelKey(datasetId: string, modelId: string) {
  return `${datasetId}:${modelId}`;
}

export function createBigQueryModelEntity(
  data: bigquery_v2.Schema$Model,
  datasetId: string,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: getBigQueryModelKey(
          datasetId,
          data.modelReference?.modelId as string,
        ),
        _type: BIG_QUERY_MODEL_ENTITY_TYPE,
        _class: BIG_QUERY_MODEL_ENTITY_CLASS,
        name: data.modelReference?.modelId,
        displayName: data.friendlyName!,
        description: data.description,
        etag: data.etag,
        modelType: data.modelType,
        location: data.location,
        createdOn: data.creationTime
          ? parseInt(data.creationTime, 10)
          : undefined,
        updatedOn: data.lastModifiedTime
          ? parseInt(data.lastModifiedTime, 10)
          : undefined,
        expirationTime: data.expirationTime
          ? parseInt(data.expirationTime, 10)
          : undefined,
        classification: null,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigquery?project=${data.modelReference?.projectId}&d=${data.modelReference?.datasetId}&m=${data.modelReference?.modelId}&p=${data.modelReference?.projectId}&page=model`,
        ),
      },
    },
  });
}

export function createBigQueryTableEntity({
  data,
  projectId,
  isPublic,
  kmsKeyName,
}: {
  data: BigQueryTable;
  projectId: string;
  isPublic: boolean | undefined;
  kmsKeyName: string | undefined;
}) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: BIG_QUERY_TABLE_ENTITY_TYPE,
        _class: BIG_QUERY_TABLE_ENTITY_CLASS,
        name: data.tableReference?.tableId,
        public: isPublic,
        type: data.type,
        friendlyName: data.friendlyName,
        classification: null,
        // 7.2 Ensure that all BigQuery Tables are encrypted with Customer-managed encryption key (CMEK)
        kmsKeyName,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigquery?project=${projectId}&d=${data.tableReference?.datasetId}&p=${projectId}&t=${data.tableReference?.tableId}&page=table`,
        ),
        createdOn: data.creationTime
          ? parseInt(data.creationTime, 10)
          : undefined,
        expirationTime: data.expirationTime
          ? parseInt(data.expirationTime, 10)
          : undefined,
      },
    },
  });
}
