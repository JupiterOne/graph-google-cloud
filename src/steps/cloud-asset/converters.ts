import { Entity } from '@jupiterone/integration-sdk-core';
import { cloudasset_v1 } from 'googleapis';
import { snakeCase } from 'lodash';
import { hashArray } from '../../utils/crypto';

import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { isReadOnlyPermission } from '../../utils/iam';
import { formatPermissions } from '../../utils/iamBindings/formatPermissions';
import { bindingEntities } from './constants';

export interface BindingEntity extends Entity {
  resource: string;
  role: string;
  members: string;
  projectId: string;
  projectName?: string;
  folders?: string[];
  organization?: string;
  'condition.title': string;
  'condition.description': string;
  'condition.expression': string;
  'condition.location': string;
  readonly: boolean;
}

export function buildIamBindingEntityKey({
  binding,
  projectName,
  resource,
}: {
  binding: cloudasset_v1.Schema$Binding;
  projectName: string | undefined | null;
  resource: string | undefined | null;
}) {
  // There are no unique values returned from this result so use everything
  // available to make the binding key unique.
  const keyBuilders = ['binding'];

  if (projectName) keyBuilders.push('project:' + projectName);
  if (resource) keyBuilders.push('resource:' + resource);
  if (binding.role) keyBuilders.push('role:' + binding.role);
  if (binding.members)
    keyBuilders.push('members:' + hashArray(binding.members));
  if (binding.condition?.expression)
    keyBuilders.push('condition:' + binding.condition.expression);

  return keyBuilders.join('|');
}

export function createIamBindingEntity({
  _key,
  binding,
  projectId,
  projectName,
  resource,
  folders,
  organization,
  permissions,
}: {
  _key: string;
  binding: cloudasset_v1.Schema$Binding;
  projectId?: string;
  projectName?: string | null;
  resource?: string | null;
  folders?: string[] | null;
  organization?: string | null;
  permissions?: string[] | null;
}): BindingEntity {
  const namePrefix = 'Role Binding for Resource: ';

  return createGoogleCloudIntegrationEntity(binding, {
    entityData: {
      source: {}, // rawData was removed to prevent `Request must be smaller than 6291456 bytes for the InvokeFunction operation` errors
      assign: {
        _class: bindingEntities.BINDINGS._class,
        _type: bindingEntities.BINDINGS._type,
        _key,
        name: snakeCase(namePrefix) + resource,
        displayName: namePrefix + resource,
        /**
         * The full resource name of the resource associated with this IAM policy.
         * Example: `//compute.googleapis.com/projects/my_project_123/zones/zone1/instances/instance1`.
         * See [Cloud Asset Inventory Resource Name Format](https://cloud.google.com/asset-inventory/docs/resource-name-format)
         * for more information. To search against the `resource`: * use a field query. Example: `resource:organizations/123`
         */
        resource,
        role: binding.role,
        members: binding.members,
        projectId,
        projectName,
        folders,
        organization,
        'condition.title': binding.condition?.title,
        'condition.description': binding.condition?.description,
        'condition.expression': binding.condition?.expression,
        'condition.location': binding.condition?.location,
        readonly: permissions?.some((p) => !isReadOnlyPermission(p)) ?? true, // default to true if there are no permissions
        ...formatPermissions(permissions ?? []),
      },
    },
  }) as BindingEntity;
}
