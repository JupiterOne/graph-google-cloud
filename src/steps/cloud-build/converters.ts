import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
import { cloudbuild_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { CloudBuildEntitiesSpec } from './constants';

export const createGoogleCloudBuildEntity = (
  source: cloudbuild_v1.Schema$Build,
) => {
  const data = {
    ...source,
    labels: source.tags,
    tags: undefined,
  };

  return createGoogleCloudIntegrationEntity(source, {
    entityData: {
      source: source,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD._class,
        _type: CloudBuildEntitiesSpec.BUILD._type,
        _key: data.id as string,
        name: data.id as string,
        id: data.id as string,
        status: data.status,
        'source.storageSource.bucket': data.source?.storageSource?.bucket,
        'source.storageSource.object': data.source?.storageSource?.object,
        'source.storageSource.generation':
          data.source?.storageSource?.generation,
        'results.buildStepImages': data.results?.buildStepImages,
        'results.buildStepOutputs': data.results?.buildStepOutputs,
        createTime: parseTimePropertyValue(data.createTime),
        startTime: parseTimePropertyValue(data.startTime),
        finishTime: parseTimePropertyValue(data.finishTime),
        timeout: data.timeout,
        projectId: data.projectId,
        logsBucket: data.logsBucket,
        'options.logging': data.options?.logging,
        'options.pool': data.options?.workerPool,
        logUrl: data.logUrl,
        queueTtl: data.queueTtl,
      },
    },
  });
};

export const createGoogleCloudBuildTriggerEntity = (
  source: cloudbuild_v1.Schema$BuildTrigger,
) => {
  const data = {
    ...source,
    labels: source.tags,
    tags: undefined,
  };

  return createGoogleCloudIntegrationEntity(source, {
    entityData: {
      source: source,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_TRIGGER._class,
        _type: CloudBuildEntitiesSpec.BUILD_TRIGGER._type,
        _key: data.id as string,
        name: data.name,
        id: data.id as string,
        createdOn: parseTimePropertyValue(source.createTime),
      },
    },
  });
};

export const createGoogleCloudBuildWorkerPoolEntity = (
  data: cloudbuild_v1.Schema$WorkerPool,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: {
        ...data,
        tags: data.annotations,
      },
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_WORKER_POOL._class,
        _type: CloudBuildEntitiesSpec.BUILD_WORKER_POOL._type,
        _key: data.uid as string,
        name: data.name,
        displayName: data.displayName as string,
        createdOn: parseTimePropertyValue(data.createTime),
        updatedOn: parseTimePropertyValue(data.updateTime),
        deletedOn: parseTimePropertyValue(data.deleteTime),
        state: data.state,
        uid: data.uid,
      },
    },
  });
};

export const createGoogleCloudBuildGithubEnterpriseConfigEntity = (
  data: cloudbuild_v1.Schema$GitHubEnterpriseConfig,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_GITHUB_ENTERPRISE_CONFIG._class,
        _type: CloudBuildEntitiesSpec.BUILD_GITHUB_ENTERPRISE_CONFIG._type,
        _key: data.name as string,
        name: data.name,
        displayName: data.displayName as string,
        appId: data.appId,
        createdOn: parseTimePropertyValue(data.createTime),
        hostUrl: data.hostUrl,
        peeredNetwork: data.peeredNetwork,
        'secrets.oauthClientIdVersionName':
          data.secrets?.oauthClientIdVersionName,
        'secrets.oauthSecretVersionName': data.secrets?.oauthSecretVersionName,
        'secrets.privateKeyVersionName': data.secrets?.privateKeyVersionName,
        'secrets.webhookSecretVersionName':
          data.secrets?.webhookSecretVersionName,
        sslCa: data.sslCa,
        webhookKey: data.webhookKey,
      },
    },
  });
};

export const createGoogleCloudBuildBitbucketServerConfigEntity = (
  data: cloudbuild_v1.Schema$BitbucketServerConfig,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_BITBUCKET_SERVER_CONFIG._class,
        _type: CloudBuildEntitiesSpec.BUILD_BITBUCKET_SERVER_CONFIG._type,
        _key: data.name as string,
        name: data.name,
        hostUri: data.hostUri,
        createTime: parseTimePropertyValue(data.createTime),
        username: data.username,
        webhookKey: data.webhookKey,
        apiKey: data.apiKey,
        'secrets.adminAccessTokenVersionName':
          data?.secrets?.adminAccessTokenVersionName,
        'secrets.readAccessTokenVersionName':
          data?.secrets?.readAccessTokenVersionName,
        'secrets.webhookSecretVersionName':
          data?.secrets?.webhookSecretVersionName,
      },
    },
  });
};

export const createGoogleCloudBuildBitbucketRepoEntity = (
  data: cloudbuild_v1.Schema$BitbucketServerRepository,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_BITBUCKET_REPO._class,
        _type: CloudBuildEntitiesSpec.BUILD_BITBUCKET_REPO._type,
        _key: data.name as string,
        name: data.name,
        browseUri: data.browseUri,
        description: data.description,
        displayName: data.displayName as string,
        'repoId.projectKey': data.repoId?.projectKey,
        'repoId.repoSlug': data.repoId?.repoSlug,
        'repoId.webhookId': data.repoId?.webhookId,
      },
    },
  });
};
