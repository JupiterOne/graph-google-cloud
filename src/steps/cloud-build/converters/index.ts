import google_cloud_build from './google_cloud_build';
import google_cloud_build_trigger from './google_cloud_build_trigger';

export default {
  google_cloud_build: google_cloud_build,
  google_cloud_build_trigger: google_cloud_build_trigger,
};
