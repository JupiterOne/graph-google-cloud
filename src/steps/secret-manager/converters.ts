import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
import { secretmanager_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { SecretManagerEntities } from './constants';

export function createSecretEntity(data: secretmanager_v1.Schema$Secret) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: SecretManagerEntities.SECRET._class,
        _type: SecretManagerEntities.SECRET._type,
        _key: data.name as string,
        name: data.name,
        createdAt: parseTimePropertyValue(data.createTime),
        expiresAt: parseTimePropertyValue(data.expireTime),
        automaticReplicationKmsKeyName:
          data.replication?.automatic?.customerManagedEncryption?.kmsKeyName,
        nextRotationTime: parseTimePropertyValue(
          data.rotation?.nextRotationTime,
        ),
        rotationPeriod: data.rotation?.rotationPeriod,
        topicNames: data.topics?.map((t) => t.name!),
        ttl: data.ttl,
      },
    },
  });
}

export function createSecretVersionEntity(
  data: secretmanager_v1.Schema$SecretVersion,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: SecretManagerEntities.SECRET_VERSION._class,
        _type: SecretManagerEntities.SECRET_VERSION._type,
        _key: data.name as string,
        name: data.name,
        createdAt: parseTimePropertyValue(data.createTime),
        destroyTime: parseTimePropertyValue(data.destroyTime),
      },
    },
  });
}
