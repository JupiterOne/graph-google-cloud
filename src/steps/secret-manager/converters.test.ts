import {
  getMockSecretEntity,
  getMockSecretVersionEntity,
} from '../../../test/mocks';
import { createSecretEntity } from './converters';

describe('#createSecretEntity', () => {
  test('should convert to entity', () => {
    expect(createSecretEntity(getMockSecretEntity())).toMatchSnapshot();
  });
});

describe('#createSecretVersionEntity', () => {
  test('should convert to entity', () => {
    expect(createSecretEntity(getMockSecretVersionEntity())).toMatchSnapshot();
  });
});
