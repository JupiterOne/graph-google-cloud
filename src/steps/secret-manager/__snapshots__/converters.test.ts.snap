// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createSecretEntity should convert to entity 1`] = `
{
  "_class": [
    "Group",
  ],
  "_key": "sample-secret-name",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "createTime": "2022-08-10T13:39:10.097472Z",
        "name": "sample-secret-name",
      },
    },
  ],
  "_type": "google_secret_manager_secret",
  "automaticReplicationKmsKeyName": undefined,
  "createdAt": 1660138750097,
  "createdOn": undefined,
  "displayName": "sample-secret-name",
  "expiresAt": undefined,
  "name": "sample-secret-name",
  "nextRotationTime": undefined,
  "rotationPeriod": undefined,
  "topicNames": undefined,
  "ttl": undefined,
}
`;

exports[`#createSecretVersionEntity should convert to entity 1`] = `
{
  "_class": [
    "Group",
  ],
  "_key": "projects/j1-gc-integration-dev-v3/secrets/sample-secret-name/versions/1",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "createTime": "2022-08-10T14:19:20.840979Z",
        "name": "projects/j1-gc-integration-dev-v3/secrets/sample-secret-name/versions/1",
        "state": "ENABLED",
      },
    },
  ],
  "_type": "google_secret_manager_secret",
  "automaticReplicationKmsKeyName": undefined,
  "createdAt": 1660141160840,
  "createdOn": undefined,
  "displayName": "projects/j1-gc-integration-dev-v3/secrets/sample-secret-name/versions/1",
  "expiresAt": undefined,
  "name": "projects/j1-gc-integration-dev-v3/secrets/sample-secret-name/versions/1",
  "nextRotationTime": undefined,
  "rotationPeriod": undefined,
  "topicNames": undefined,
  "ttl": undefined,
}
`;
